interfaces/vault_kv/v0/README.md

```diff
@@ -0,0 +1,67 @@
+# `vault-kv`
+
+## Usage
+
+Some charms require a secure key value store. This relation interface describes the expected behavior of any charm claiming to interact with Vault Key Value stores.
+
+## Direction
+
+```mermaid
+flowchart TD
+    Requirer -- secret_backend, egress_subnet --> Provider
+    Provider -- vault_url, kv_mountpoint, credentials --> Requirer
+```
+
+## Behavior
+
+Both the Requirer and the Provider need to adhere to criteria to be considered compatible with the interface.
+
+### Provider
+
+Provider expectations
+
+- Must provide the vault url
+- Must provide a key value mountpoint
+- Must provide a role_id and role_secret_id for each unit, with access protected by unit's egress_subnet
+
+### Requirer
+
+Requirer expectations
+
+- Must provide a secret backend name, which must start with "charm-"
```

Comment:
I think we might want a way for the requirer to provide a part of the name, in the case where multiple secret backends are required. A unique ID with a different relation might work, but it would be more difficult for admins to identify at a glance.

Reply:
Yes, the goal is to be able to have multiple relations with a different secret backend.

Reply:
What would be a better approach?

`<model>-<app>-<user specified suffix>`?

Reply:
The vault charm is currently creating the following rules:
```hcl
# Allow management of policies starting with charm- prefix
path "sys/policy/charm-*" {
  capabilities = ["create", "read", "update", "delete"]
}

# Allow management of approle's with charm- prefix
path "auth/approle/role/charm-*" {
  capabilities = ["create", "read", "update", "delete", "list"]
}

# Allow charm- prefixes secrets backends to be mounted and managed
path "sys/mounts/charm-*" {
  capabilities = ["create", "read", "update", "delete", "sudo"]
}
```

This is where the `charm-` requirement comes from.

If we want to take the model name, should it be the model Vault is in, or the model the app is in ? (in case of a CMR) Seems like it would make more sense to use the model the requirer app is in following your example

Reply:
The vault charm won't have any way to know about the requirer model, so that can't be the case. I would be fine with `<charm>-<requirer app>-<user specified suffix>`

Reply:
There's still the edge case when a requiring application is deployed twice in different models (same app name)

Reply:
Good point, should we also add a new suffix for a unique id? `<charm>-<requirer app>-<user suffix>-<uuid>`.

Overall this feels like way too long endpoints.

Is there an issue with those mounts not being human readable? If the mountpoints are simply generated by the vault charm and provided back to the unit ?

@ghislainbourgeois Do you have any opinion on this?

Reply:
There's an issue, what's the required behavior if we break the relationship ? Should the KV backend be erased ? Or should a new integration of the same app connect back to the same KV ?

If there's a custom UUID generated on connection created/join you'll never be able to access back the data from the previous relation, and I'm not sure that's what we want.


About human readable, this only applies if we want users to easily use (and not copy paste) the vault CLI to access the mount, which is out of scope of a Juju relationship I think

Reply:
I think that if the relation is broken (meaning that there is a user intent behind it, not jut a pod going down) the KV backend should be disabled. If not, this could lead to security issues no? I guess we should align with what other charms do (ex. db's).

Reply:
Looking at [postgresql-k8s-operator](https://github.com/canonical/postgresql-k8s-operator/blob/main/src/relations/postgresql_provider.py), this charm deletes the relation user, but keeps the database. When a new relation requests the same database, a new relation user is created and given rights on the database.

Reply:
What could be done to mimic the behavior of db charms is to disable the secret engine, which will revoke every secret leases. (lease should hold no effect over as kv backend do emit leases [docs](https://developer.hashicorp.com/vault/docs/v1.10.x/concepts/lease).

I will confirm if re-enabling the secrets engine keeps stored secrets available

Reply:
Moreover, from juju [docs](https://juju.is/docs/juju/manage-cross-model-integrations#heading--suspend-resume-or-remove-an-integration), using `juju suspend-relation` will run the departed/broken hooks, and if we disable the secrets engine, using `juju resume-relation` would not actually resume, since the secrets data would be removed.

It looks like we cannot differ `juju remove-relation` from `juju suspend-relation` from the charm

Reply:
I am not sure the user intent behind removing the relation can be equated to the user wanting to delete the data. I would be very wary of deleting data automatically and assuming user intent.

Additionally, some deployments might have data retention requirements that this would break.

If we want to provide cleanup, I would prefer to expose actions, or expose the UI or CLI for the admin.

Regarding the naming convention, I think the requirer app name in cross-model relations is anonymized, so `<charm>-<requirer app>-<user specified suffix>` would most likely work in all cases.

Reply:
I agree with your analysis, I don't believe we should delete the data.

The requirer app name still works with app name being `remote-<some uuid>`.

I'm currently facing this [bug](https://bugs.launchpad.net/juju/+bug/2033486) when trying to share the secrets over CMR.

Reply:
I think this is good enough, the provided suffix would help the admin identify the requirer if it is well chosen, and I do not think there is much more we can do in the interface to enforce that.